api/exec_start.go

```diff
@@ -0,0 +1,217 @@
+package api
+
+import (
+	"bytes"
+	"context"
+	"encoding/binary"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"time"
+
+	"github.com/hashicorp/nomad/plugins/drivers"
+)
+
+type ExecStartRequest struct {
+
+	// streams
+	Stdin  io.Reader
+	Stdout io.Writer
+	Stderr io.Writer
+
+	// terminal size channel
+	ResizeCh <-chan drivers.TerminalSize
+
+	// Tty indicates whether pseudo-terminal is to be allocated
+	Tty bool
+
+	// AttachOutput is whether to attach to STDOUT
+	// If false, stdout will not be attached
+	AttachOutput bool
+	// AttachError is whether to attach to STDERR
+	// If false, stdout will not be attached
+	AttachError bool
+	// AttachInput is whether to attach to STDIN
+	// If false, stdout will not be attached
+	AttachInput bool
+}
+
+// DemuxHeader reads header for stream from server multiplexed stdin/stdout/stderr/2nd error channel
+func DemuxHeader(r io.Reader, buffer []byte) (fd, sz int, err error) {
+	n, err := io.ReadFull(r, buffer[0:8])
+	if err != nil {
+		return
+	}
+	if n < 8 {
+		err = io.ErrUnexpectedEOF
+		return
+	}
+
+	fd = int(buffer[0])
+	if fd < 0 || fd > 3 {
+		err = fmt.Errorf(`channel "%d" found, 0-3 supported`, fd)
+		return
+	}
+
+	sz = int(binary.BigEndian.Uint32(buffer[4:8]))
+	return
+}
+
+// DemuxFrame reads contents for frame from server multiplexed stdin/stdout/stderr/2nd error channel
+func DemuxFrame(r io.Reader, buffer []byte, length int) (frame []byte, err error) {
+	if len(buffer) < length {
+		buffer = append(buffer, make([]byte, length-len(buffer)+1)...)
+	}
+
+	n, err := io.ReadFull(r, buffer[0:length])
+	if err != nil {
+		return nil, nil
+	}
+	if n < length {
+		err = io.ErrUnexpectedEOF
+		return
+	}
+
+	return buffer[0:length], nil
+}
+
+// This is intended to be run as a goroutine, handling resizing for a container
+// or exec session.
+func (c *API) attachHandleResize(ctx context.Context, resizeChannel <-chan drivers.TerminalSize, sessionId string) {
+	for {
+		select {
+		case <-ctx.Done():
+			c.logger.Warn("Resize handler is done")
+			return
+		case size := <-resizeChannel:
+			c.logger.Debug("Resize terminal", "sessionId", sessionId, "height", size.Height, "width", size.Width)
+			rerr := c.ExecResize(ctx, sessionId, size.Height, size.Width)
+			if rerr != nil {
+				c.logger.Warn("failed to resize TTY", "err", rerr)
+			}
+		}
+	}
+}
+
+// ExecStartAndAttach starts and attaches to a given exec session.
+func (c *API) ExecStart(ctx context.Context, sessionID string, options ExecStartRequest) error {
+	client := *c.httpClient
```

Comment:
having our own clone function may be helpful here, but for now I think something like the following would be a bit safer
```
	client := new(http.Client)
	*client = *c.httpClient
```

Reply:
I'm fine with the change because i am not really a go expert ;-)